### DISCUSSION TITLE
Ref stays undefined if accessed before the element was added to the document

### DISCUSSION THREAD

**Discussion by alexandergitter** [alpinejs/alpine#2713] [7. Bugs (archived)] [answered]
Hi,

take a look at the following example: https://jsfiddle.net/328xa6te/

When you click the "Add Ref element" button, a new div gets added with an `x-ref` attribute. Clicking on "Log Ref" correctly logs the element to the console.

Doing it the other way around however does not work:
- Click "Log Ref" and observe `undefined` getting logged
- Click "Add Ref element"
- Click "Log Ref" again - still `undefined`, despite the `x-ref` now being present in the document

Is there a way to make Alpine pick up changed x-ref elements or is this a bug?

**Comment by SimoTod** [comment]
refs are cached the first time you do a lookup, they can't be changed dynamically. This is by design as far as I know: https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/magics/%24refs.js#L6

**Reply by SimoTod** [reply]
you can work around it doing `<button @click="delete $el._x_refs_proxy; console.log($refs.myref)">Log Ref</button>` if you really need but you'll be using a private internal api that could change any time in the future

**Reply by alexandergitter** [reply]
Mhh... it feels a bit inconsistent though, since the following _does_ appear to work as expected: https://jsfiddle.net/rdutLqjo/

This is just adding an `x-ref` to the parent element (the alpine component itself in this case)

**Reply by SimoTod** [reply]
I can't have a look right now because jsfiddle doesn't work on mobile but, in general, $refs creates a local cache the first time you lookup for a reference. Other elements won't use that cache and if you lookup when the element is already in page it does work.
That being said, it doesn't have to work like that it's just the current internal implementation so, if you have a good implementation that covers your use case, you can try to send a PR.

**Reply by alexandergitter** [reply]
See https://github.com/alpinejs/alpine/discussions/2713#discussioncomment-2299228

**Comment by alexandergitter** [comment, marked as answer]
I now use something like the following, which seems to work fine for my use case:

```js
Alpine.magic('lazyref', (el, { Alpine }) =>
  new Proxy({}, {
    get(target, ref) {
      return Alpine.closestRoot(el).querySelector(`[x-ref="${ref}"]`)
    }
  })
)
```
